crates/file_formats/dat/src/tileedge/mod.rs

```diff
@@ -0,0 +1,81 @@
+// OpenAOE: An open source reimplementation of Age of Empires (1997)
+// Copyright (c) 2016 Kevin Fuller
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::{SeekFrom};
+use std::path::Path;
+use error::*;
+
+use io_tools::*;
+
+#[derive(Default, Debug)]
+pub struct OutlineEntry {
+    pub index: u8,
+    pub left_offset: u8,
+    pub right_offset: u8
+}
+
+#[derive(Default, Debug)]
+pub struct TileEdgeEntry {
+    pub outlines: Vec<OutlineEntry>
+}
+
+#[derive(Default, Debug)]
+pub struct TileEdgeDb {
+    pub entries: Vec<TileEdgeEntry>
+}
+
+impl TileEdgeDb {
+    fn new() -> TileEdgeDb {
+        Default::default()
+    }
+
+    pub fn read_from_file<P: AsRef<Path>>(file_name: P) -> Result<TileEdgeDb> {
+        let mut file = try!(File::open(file_name.as_ref()));
+        file.seek(SeekFrom::Start(444));
+
+        let mut db = TileEdgeDb::new();
+
+        // 1586 is the biggest number I could find that wouldn't make the reader crash because it reached EOF
+        'entry: for _ in 1..1586 {

```

Comment:
I think it's probably better to read the entire file into a Vec so that you know exactly how many outlines you're dealing with:

```
let mut bytes = Vec::new();
try!(file.read_to_end(&mut bytes));

for _ in 0..(bytes.len() / 3) {
   ...
} 
```